lib/nexthop.h

```diff
@@ -116,15 +118,26 @@ struct nexthop {
 
 	/* Weight of the nexthop ( for unequal cost ECMP ) */
 	uint8_t weight;
+
+	/* Index of backup nexthop in a backup list, if valid */
+	uint8_t backup_idx;
```

Comment:
I don't understand the purpose of the index here. Can you elaborate a bit?

Its to match and index to another nexthop in the backup_group? Like a 1:1 relationship?

Reply:
As I wrote above, I've tried to describe this indexing approach in the design doc I linked in the PR. There's a N:1 relationship possible between one or more primary nexthops and a backup nexthop. Of course that can be 1:1, if that's what the sender/source wants to do, and it may be that some primaries don't have viable backups at all. All of that should be possible. What's not possible is for a single primary nexthop to have multiple backups.

Reply:
Hmmm okay I see.

So question, from the code it looked like your intent was to add backup nexthops into their own `dependency` RB tree eventually like is done with other nexthops.

In those trees there is no guarantee on order so using an `index` value doesn't really make sense. That is really only useful in the linked list lib/nexthop.

Once you start talking about the NHE dependency trees, ordering of a group is moot. The order is determined by the ID since that is what the RB tree uses as its compare function.

So I am not sure how backups can translate into being NHE objects with this approach.

Reply:
yes, the indexing only applies to the ordered backup list. I'm sort of deferring the netlink-specific aspect until we (some netlink folks) decide what, if anything, should be done. for now, the backups are segregated, and in the netlink nexthop-id path they are skipped. the netlink semantics are different, and it's not clear how (or whether) this concept will apply.

Reply:
Sure, but if we did decide to include backup nexthops in the NHE dependency tree eventually and get netlink to use them, I feel it would require us to rewrite a lot of this since its dependent on ordering of an index to work.

Why not go ahead and plan on netlink people wanting to use backup nexthops from the getgo?

My initial though was if we moved setting IDs into the nexthop config and into the nexthop struct somehow, you might could just reference them by ID.

What do you think about that? Or is the ordering important for something other unknown aspect I am not aware of?

Reply:
I think (think) the netlink issue is more fundamental: there's no concept of backing up a nexthop, at all. What might be possible, and what seems to be done manually sometimes, would be to create a second instance of a route, using a less-preferred metric. The semantics and behavior there are ... just different, but it would be possible. In any case, I'd prefer not to try to imagine what _might_ happen in netlink, eventually. I'd like to see about getting backup nexthops flowing, then when and if something emerges from the netlink community, we can figure out how to take advantage of that.

I don't really quite get the problem for zebra, I guess. We have a route with an nhg/nhe, and we have the lists of primary and backup nexthops. A forwarding plane can use the relationship in route updates because the messages contain those lists. And we could do the 'less-preferred route' approach for netlink with a minimal change - again, I think - if we/you/they decide that that's desirable. All I'm avoiding here - initially - is the work of installing netlink nexthops when they aren't being used.

Reply:
So you know my plan (pending your alternative approach we will hear later) is to continue developing the NHE dependency trees add adding features to them to quickly validate/de-validate nexthops based on interface events and such.

If the backup nexthops arent included in those dependency trees somehow (and I dont think they can be if they depend on index ordering), they aren't going to get those features or any other related to the dependency trees. See my concern?

Sorry not trying to be abrasive, just voicing my concerns.

Reply:
Gosh no - nothing abrasive about it! Currently, the backup nexthops _do_ get resolved into trees of nhes, and they get processed just as the primary nexthops' nhes do. The only thing they don't get is installed, by id, in netlink, because it's not clear how to make use of that, yet. We may decide we want to distinguish primary and backup nhes, and then we could process them to do a reference by nexthop-id (which should be stable and unique), and we could set that up easily enough. I've just deferred doing that because it's not clear that that is the right direction for netlink right now. I do consider myself committed to making (or helping make) backups work for netlink, when / if we figure out how to approach that.

Reply:
oh okay maybe i didnt follow the code properly. good to me then.